Add documentations for MLflow authentication (#8626)
mlflow/mlflow: docs/source/auth/index.rst

> .. _auth:

Reviewer: This looks fantastic!

Reviewer: I'm going to file a PR to allow setting up credentials in `~/.mlflow/credentials` :) We might need a small adjustment once that's merged.

Reviewer: Opened #8688

Author: thanks! :) I'll add the missing part

In Python, you can use the ``requests`` library:

.. code-block:: python

    import requests

    response = requests.get(
        "https://mlflow_tracking.uri/",
        auth=("username", "password"),
    )

Using MLflow Client
-------------------

MLflow provides two environment variables for authentication: ``MLFLOW_TRACKING_USERNAME`` and ``MLFLOW_TRACKING_PASSWORD``.
To use basic authentication, you must set both environment variables.

.. code-block:: python

    import os
    import mlflow

    os.environ["MLFLOW_TRACKING_USERNAME"] = "username"
    os.environ["MLFLOW_TRACKING_PASSWORD"] = "password"

    mlflow.set_tracking_uri("https://mlflow_tracking.uri/")
    with mlflow.start_run():
        ...

Creating New User
=================

Using MLflow UI
---------------

MLflow UI provides a simple page for creating new users at ``<tracking_uri>/signup``.

    .. image:: ../_static/images/auth_signup.png

Using REST API
--------------

Alternatively, you can send ``POST`` requests to the Tracking Server endpoint ``2.0/users/create``.

In Python, you can use the ``requests`` library:

.. code-block:: python

    import requests

    response = requests.post(

>         "https://mlflow_tracking.uri/api/2.0/mlflow/users/create",

Reviewer: ```suggestion
        "https://<mlflow_tracking_uri>/api/2.0/mlflow/users/create",
```

Can we add `<>` here to clarify this is a placeholder that needs to be replaced?

Author: sure!

    mlflow server --app-name basic-auth

Server admin can choose to disable this feature anytime by restarting the server without the ``app-name`` flag. 
Any users and permissions created will be persisted on a SQL database and will be back in service once the feature is re-enabled.

Due to the nature of HTTP authentication, it is only supported on a remote Tracking Server, where users send
requests to the server via REST APIs.

How It Works
============

Permissions
-----------

The available permissions are:

.. list-table::
   :widths: 10 10 10 10 10
   :header-rows: 1

   * - Permission
     - Can read
     - Can update
     - Can delete
     - Can manage
   * - ``READ``
     - Yes
     - No
     - No
     - No
   * - ``EDIT``
     - Yes
     - Yes
     - No
     - No
   * - ``MANAGE``
     - Yes
     - Yes
     - Yes
     - Yes
   * - ``NO_PERMISSIONS``
     - No
     - No
     - No
     - No

The default permission for all users is ``READ``. It can be changed in the :ref:`configuration <configuration>` file.

Permissions can be set on two MLflow components: Experiments and Registered Models.
Certain APIs can only be requested if the user has the required permission.

> Otherwise, a "Permission Denied" response will be returned.

Author: sure

.. list-table::
   :widths: 10 10 10 10
   :header-rows: 1

   * - API
     - Endpoint
     - Method

>      - Permission

Reviewer: ```suggestion
     - Required permission
```

for clarity.

   * - :ref:`Create Experiment <mlflowMlflowServicecreateExperiment>`
     - ``2.0/mlflow/experiments/create``
     - ``POST``
     - None
   * - :ref:`Get Experiment <mlflowMlflowServicegetExperiment>`
     - ``2.0/mlflow/experiments/get``
     - ``GET``

>      - ExperimentPermission.can_read

Reviewer: If we split the table, can we remove `ExperimentPermission`?

Author: i have changed the section a bit

     - ``GET``
     - Only returns experiments which the user has ``READ`` permission on.
   * - :ref:`Search Runs <mlflowMlflowServicesearchRuns>`
     - ``2.0/mlflow/runs/search``
     - ``POST``
     - Only returns experiments which the user has ``READ`` permission on.
   * - :ref:`Search Registered Models <mlflowModelRegistryServicesearchRegisteredModels>`
     - ``2.0/mlflow/registered-models/search``
     - ``GET``
     - Only returns registered models which the user has ``READ`` permission on.
   * - :ref:`Search Model Versions <mlflowModelRegistryServicesearchModelVersions>`
     - ``2.0/mlflow/model-versions/search``
     - ``GET``
     - Only returns registered models which the user has ``READ`` permission on.

Permissions Database
--------------------

All users and permissions are stored in a database in ``basic_auth.db``, relative to the directory where MLflow server is launched.
The location can be change in the :ref:`configuration <configuration>` file.

Admin Users
-----------

Admin users have unrestricted access to all MLflow resources,
**including creating or deleting users, updating password and admin status of other users,

> grating or revoking permissions from other users, and managing permissions for all 

Reviewer 2: ```suggestion
granting or revoking permissions from other users, and managing permissions for all 
```

> MLflow resources,** even if ``NO_PERMISSIONS`` is explicitly set to an admin.

Reviewer 2: ```suggestion
MLflow resources,** even if ``NO_PERMISSIONS`` is explicitly set to that admin account.
```

> MLflow has a built-in admin user that will be created the first time MLflow authentication feature is enabled.

Reviewer 2: ```suggestion
MLflow has a built-in admin user that will be created the first time that the MLflow authentication feature is enabled.
```

> The default admin user credentials are as follows.
> It is recommended that you update the password as soon as possible after it is created.

Reviewer 2: ```suggestion

.. note::
    It is recommended that you update the password as soon as possible after it is created.

The default admin user credentials are as follows:

```

:py:func:`AuthServiceClient<mlflow.server.auth.client.AuthServiceClient>` to manage users and permissions.
To instantiate ``AuthServiceClient``, it is recommended that you use :py:func:`mlflow.server.get_app_client`.

.. code-block:: python
    :caption: Example

    import os
    from mlflow import MlflowClient
    from mlflow.server import get_app_client

    tracking_uri = "http://localhost:5000/"

    auth_client = get_app_client("basic-auth", tracking_uri=tracking_uri)
    auth_client.create_user(username="user1", password="pw1")
    auth_client.create_user(username="user2", password="pw2")

    os.environ["MLFLOW_TRACKING_USERNAME"] = "user1"
    os.environ["MLFLOW_TRACKING_PASSWORD"] = "pw1"

    client = MlflowClient(tracking_uri=tracking_uri)
    experiment_id = client.create_experiment(name="experiment")

    auth_client.create_experiment_permission(
        experiment_id=experiment_id, username="user2", permission="MANAGE"
    )

Authenticating to MLflow
========================

Using MLflow UI
---------------

When a user first visits the MLflow UI on a browser, they will be prompted to login. 
There is no limit to how many login attempts can be made.

> Currently, MLflow UI does not display any information of the current user.

Reviewer 2: ```suggestion
Currently, MLflow UI does not display any information about the current user.
```

Once a user is logged in, the only way to log out is to close the browser.

    .. image:: ../_static/images/auth_prompt.png

> Using REST API

Reviewer: Can we modify the order of sub-sections in this section?

- Environment variables
- Credentials File
- REST API

Author: okay

--------------

A user can authenticate using the HTTP ``Authorization`` request header.
See https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication for more information.

In Python, you can use the ``requests`` library:

.. code-block:: python

    import requests

    response = requests.get(
        "https://<mlflow_tracking_uri>/",
        auth=("username", "password"),
    )

Using MLflow Client
-------------------

MLflow provides two environment variables for authentication: ``MLFLOW_TRACKING_USERNAME`` and ``MLFLOW_TRACKING_PASSWORD``.
To use basic authentication, you must set both environment variables.

.. code-block:: python

    import os
    import mlflow

    os.environ["MLFLOW_TRACKING_USERNAME"] = "username"

>     os.environ["MLFLOW_TRACKING_PASSWORD"] = "password"

Author: updated

    mlflow.set_tracking_uri("https://<mlflow_tracking_uri>/")
    with mlflow.start_run():
        ...

Using Credentials File
----------------------

> You can save your credentials in a file to remove the need of setting environment variables every time.

Reviewer 2: ```suggestion
You can save your credentials in a file to remove the need for setting environment variables every time.
```

> The credentials should be saved in ``~/.mlflow/credentials`` using INI format. Note that the password 

Reviewer 2: ```suggestion
The credentials should be saved in ``~/.mlflow/credentials`` using ``INI`` format. Note that the password 
```

will be stored unencrypted on disk, and is protected only by filesystem permissions.

If the environment variables ``MLFLOW_TRACKING_USERNAME`` and ``MLFLOW_TRACKING_PASSWORD`` are configured,
they override any credentials provided in the credentials file.

.. code-block:: ini
    :caption: Credentials file format

    [mlflow]
    mlflow_tracking_username = username
    mlflow_tracking_password = password

> Creating New User

Reviewer 2: ```suggestion
Creating a New User
```
